File: ditto-go-client/atomic_errors_contract_test.go
package ditto

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type atomicErrorContractSuite struct {
	Cases []atomicErrorContractCase `json:"cases"`
}

type atomicErrorContractCase struct {
	ID        string                    `json:"id"`
	Operation string                    `json:"operation"`
	Inputs    atomicErrorContractInputs `json:"inputs"`
	Expect    atomicErrorContractExpect `json:"expect"`
}

type atomicErrorContractInputs struct {
	Transport     string `json:"transport"`
	Status        int    `json:"status"`
	OperationName string `json:"operation_name"`
	Body          string `json:"body"`
	ErrorMessage  string `json:"error_message"`
	ErrorKind     string `json:"error_kind"`
	ErrorCode     string `json:"error_code"`
}

type atomicErrorContractExpect struct {
	Code            string `json:"code"`
	MessageContains string `json:"message_contains"`
}

func TestAtomicErrorsContract(t *testing.T) {
	contractPath := filepath.Join("..", "contracts", "atomic-errors.contract.json")
	raw, err := os.ReadFile(contractPath)
	if err != nil {
		t.Fatalf("read contract: %v", err)
	}

	var suite atomicErrorContractSuite
	if err := json.Unmarshal(raw, &suite); err != nil {
		t.Fatalf("parse contract: %v", err)
	}

	for _, tc := range suite.Cases {
		tc := tc
		t.Run(tc.ID, func(t *testing.T) {
			switch tc.Operation {
			case "normalize_http_atomic_error":
				assertAtomicDittoError(t, parseAtomicHTTPUnsupported([]byte(tc.Inputs.Body), tc.Inputs.OperationName), tc.Expect)
			case "normalize_tcp_atomic_error":
				var input error
				if tc.Inputs.ErrorKind == "ditto" {
					input = &DittoError{Code: tc.Inputs.ErrorCode, Message: tc.Inputs.ErrorMessage}
				} else {
					input = errors.New(tc.Inputs.ErrorMessage)
				}
				assertAtomicDittoError(t, normalizeAtomicTCPErr(input, tc.Inputs.OperationName), tc.Expect)
			default:
				t.Fatalf("unsupported contract operation: %s", tc.Operation)
			}
		})
	}
}

func assertAtomicDittoError(t *testing.T, err error, expect atomicErrorContractExpect) {
	t.Helper()
	de, ok := err.(*DittoError)
	if !ok {
		t.Fatalf("expected DittoError, got %T (%v)", err, err)
	}
	if de.Code != expect.Code {
		t.Fatalf("expected code %q, got %q", expect.Code, de.Code)
	}
	if !strings.Contains(de.Message, expect.MessageContains) {
		t.Fatalf("expected message containing %q, got %q", expect.MessageContains, de.Message)
	}
}

File: ditto-nodejs-client/tests/atomic-errors-contract.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import {
  normalizeAtomicUnsupportedError,
  unsupportedAtomicHttpError,
} from '../dist/client-internal.js';
import { DittoError } from '../dist/types.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const contractPath = path.resolve(__dirname, '../../contracts/atomic-errors.contract.json');
const contract = JSON.parse(fs.readFileSync(contractPath, 'utf8'));

test('atomic error normalization contract (node shared helpers)', async () => {
  for (const c of contract.cases) {
    if (c.operation === 'normalize_http_atomic_error') {
      const response = new Response(c.inputs.body, {
        status: c.inputs.status,
        headers: { 'content-type': 'application/json' },
      });
      const err = await unsupportedAtomicHttpError(response, c.inputs.operation_name);
      assertAtomicError(err, c.expect);
      continue;
    }

    if (c.operation === 'normalize_tcp_atomic_error') {
      const input = c.inputs.error_kind === 'ditto'
        ? new DittoError(c.inputs.error_code, c.inputs.error_message)
        : new Error(c.inputs.error_message);
      const err = normalizeAtomicUnsupportedError(input, c.inputs.operation_name);
      assertAtomicError(err, c.expect);
      continue;
    }

    assert.fail(`Unsupported contract operation: ${c.operation}`);
  }
});

function assertAtomicError(error, expect) {
  assert.ok(error instanceof DittoError || error instanceof Error);
  assert.equal(error.code ?? error.Code, expect.code);
  assert.match(error.message, new RegExp(escapeRegExp(expect.message_contains)));
}

function escapeRegExp(value) {
  return String(value).replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
